internal/servicelog/write_buffer.go

```diff
@@ -0,0 +1,354 @@
+		return 0, io.ErrClosedPipe
+	}
+
+	writeLength := len(p)
+	if writeLength > wb.ringBuffer.Capacity() {
+		writeLength = wb.ringBuffer.Capacity()
+	}
+
+	for writeLength > wb.ringBuffer.Free() || !wb.canAdvanceHead() {
+		err := wb.releaseTail()
+		if err != nil {
+			return written, err
+		}
+	}
+
+	index := WriteIndex(atomic.LoadInt32(&wb.headIndex) + 1)
+	write := wb.getWrite(index)
+	write.index = index
+	write.time = writeTime
+	write.streamID = streamID
+	write.start = wb.ringBuffer.Pos()
+	n, err := wb.ringBuffer.Write(p[:writeLength])
+	if err != nil && err != io.ErrShortWrite {
+		return 0, err
+	}
+	write.end = wb.ringBuffer.Pos()
+	written = n
+
+	wb.advanceHead()
+
+	if written < len(p) {
+		return written, io.ErrShortWrite
+	}
+
+	return written, nil
+}
+
+// TailIterator returns an iterator from the tail of the stream.
+// The callee must Close the iterator when finished.
+func (wb *WriteBuffer) TailIterator() Iterator {
+	tail := wb.acquireTail()
+	return &iterator{
+		wb:    wb,
+		write: tail,
+		skip:  tail != &wb.placeholderTail,
+	}
+}
+
+// HeadIterator returns an iterator from the head of the stream.
+// If last is a positive non-zero number, the iterator will start
+// at most N writes back from the head.
+// The callee must Close the iterator when finished.
+func (wb *WriteBuffer) HeadIterator(last int) Iterator {
+	head := wb.acquireHead(last)
+	skip := head != &wb.placeholderTail
+	if last < 1 {
+		// Since the head is the last write in the buffer
+		// and the caller didn't ask for the last line, tell
+		// the iterator to not skip the first Next call.
+		skip = false
+	}
+	return &iterator{
+		wb:    wb,
+		write: head,
+		skip:  skip,
+	}
+}
+
+func (wb *WriteBuffer) getWrite(index WriteIndex) *write {
+	idx := int(index) % len(wb.writes)
+	return &wb.writes[idx]
+}
+
+func (wb *WriteBuffer) releaseTail() error {
+	wb.releaseMutex.Lock()
+	defer wb.releaseMutex.Unlock()
+	numWrites := atomic.LoadInt32(&wb.numWrites)
+	if numWrites == 0 {
+		return ErrNoMoreLines
+	}
+	for wb.placeholderTail.refs() > 0 {
+		time.Sleep(100 * time.Microsecond)
+	}
+	tailIndex := WriteIndex(atomic.LoadInt32(&wb.tailIndex))
+	l := wb.getWrite(tailIndex)
+	atomic.AddInt32(&wb.tailIndex, 1)
+	for l.refs() > 0 {
+		time.Sleep(100 * time.Microsecond)
+	}
+	if !l.empty() {
+		err := wb.ringBuffer.Release(l.start, l.end)
+		if err != nil {
+			// Release should not ever fail, but restore gracefully.
+			atomic.AddInt32(&wb.tailIndex, -1)
+			return err
+		}
+		*l = write{}
+	}
+	atomic.AddInt32(&wb.numWrites, -1)
+	return nil
+}
+
+func (wb *WriteBuffer) canAdvanceHead() bool {
+	numWrites := atomic.LoadInt32(&wb.numWrites)
+	return int(numWrites) < len(wb.writes)
+}
+
+func (wb *WriteBuffer) advanceHead() {
+	atomic.AddInt32(&wb.numWrites, 1)
+	atomic.AddInt32(&wb.headIndex, 1)
+}
+
+func (wb *WriteBuffer) signalWaiters() {
+	wb.signalMut.Lock()
+	defer wb.signalMut.Unlock()
+	close(wb.signalChan)
+	wb.signalChan = make(chan struct{})
```

Comment:
I am concerned that this is a heap allocation every log -- and in our quest to alleviate pressure on GC using ring buffers etc, we're reintroducing the allocation here anyway. I don't know how large the channel struct is (looks like it might be around 12 words or 96 bytes: https://golang.org/src/runtime/chan.go). Maybe we should do some measurements?

In any case, I don't think it's pressing -- I think as long as we have the API right, this can be considered an implementation detail.

Also, we should probably do some tests where we leave Pebble running overnight with a couple of services that log a lot, to see if there are any memory or goroutine leaks or other long-term resource issues. I can do that tomorrow.

Reply:
I'm not too concerned with this.

Next best thing would be buffered channels per iterator and a registration list. It means sending on multiple channels, but still preserves the ability of the receiver to cancel waiting (unlike with waitgroups/mutex).

Signaling becomes more expensive, but less churn on the gc.

Reply:
See benchmark results about the cost of this here: https://github.com/canonical/pebble/pull/29#issuecomment-829716115 (again, not saying we need to fix it now, just wanted to copy results while I have them handy).